---
title: "ITAS 2015"
author: "Alexander Tuzhikov"
date: "September 4, 2015"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 9
    fig_width: 12
    keep_md: yes
    number_sections: yes
    toc: yes
---


```{r libraries}
#yr libraries
source("R/yr.lib.R")
#hmptrees
source("R/hmptrees.lib.R")
#and the extension
source("R/hmptreesextend.lib.R")
#rphilyp
source("R/rphylip.lib.R")
#gbra
library("gbra")
#ggplot
source("R/ggplot2.lib.R")
#reshape2
source("R/reshape2.lib.R")
#httr
source("R/httr.lib.R")
#stringr
source("R/stringr.lib.R")
#grid
source("R/gridextra.lib.R")
#scales
source("R/scales.lib.R")
#doMC
source("R/domc.lib.R")
#doparallel
source("R/doparallel.lib.R")
#data.table
source("R/data.table.lib.R")
```

```{r qiuck prep}
source("R/helper.R")
#bh data with 4 cutoff
bh.data<- read.table(file="bh.data.minhit_4.txt")#data
legend<-tidy.legend(load.legend(legend.file = "/home/alext/Developer/gbra/inst/extradata/legend.csv",sep = "\t")) #legend
model.df<- read.table(file = "model.df.txt", header = TRUE, sep = "\t")#model descriptions
#folder helper
self.dir<- "itas.2015.output"
if(!file.exists(self.dir)){
  dir.create(self.dir)
}
self.dir.folder<- function(file=""){
  return(paste("itas.2015.output",file, sep="/"))
}
```

```{r dirty genomes}
#extract mles from the genomes
dirty.mles<- extract.gprimes(mle.list = getMLEs(df = expand.df(bh.data)))
#run fitch
rf<- Rfitch(D= dist(dirty.mles), path = "/usr/local/bin", model="FM", power=2, negative=TRUE, global=FALSE,random.order=FALSE, root=FALSE, quiet=TRUE)
#rename the tree
rf<- tidy.tree(tree = rf, legend = legend)
#save
drity.tree.newick<- "dirty.tree.newick"
write.tree(phy = rf, file = self.dir.folder(drity.tree.newick))
#plot and save results
plot.itol(newick.file = self.dir.folder(drity.tree.newick), output.file = self.dir.folder(paste0(drity.tree.newick,".pdf")))
#another one with eps, cuz i need one editable in illustrator (pdf is not in the case of iTOL)
plot.itol(newick.file = self.dir.folder(drity.tree.newick), output.file = self.dir.folder(paste0(drity.tree.newick,".eps")),format = "eps", font.size = 120)
```

```{r attach the genome ids to rsquared data}
#split names into query and target genomes
model.df.genomes<- strsplit(x = str_replace_all(string = model.df$file.name, pattern = "\\.rda", replacement = ""), split = "_", fixed = TRUE)
model.df$query.genome<- as.numeric(unlist(sapply(model.df.genomes,function(x){return(x[[1]])})))
model.df$target.genome<- as.numeric(unlist(sapply(model.df.genomes,function(x){return(x[[2]])})))
model.df$match.file.name<- str_replace_all(string = model.df$file.name, pattern = "rda", replacement = "txt.match")
```

Drosophila is **number 6** in the legend, so let's see for the rsquared distribution for this particular genome.

```{r rsquared distribution}
rsq.plot<- model.df %>% filter(query.genome==6|target.genome==6) %>% mutate(target.genome=sapply(1:length(.$target.genome), function(x){
  if(.$target.genome[x]==6){
    return(.$query.genome[x])
  }else{
    return(.$target.genome[x])
  }
})) %>% select(rsquared, target.genome) %>% merge(x=., y=legend, by.x="target.genome", by.y="id_genomes") %>%
  mutate(name=str_trim(toupper(str_replace_all(string=.$name, pattern = "\\.fasta|_|complete_genome|chromosome", replacement=" ")))) %>% 
  arrange(rsquared) %>% mutate(name=factor(.$name, levels=.$name)) %>% melt(id.vars="name") %>%
  filter(variable%in%c("rsquared")) %>% ggplot(data=., mapping=aes(x=name, y=value, group=variable, color=variable))+geom_point()+geom_smooth()+
  theme_bw()+theme(axis.text.x=element_text(angle = 45, hjust = 1),legend.position="none", plot.margin = unit(c(1, 1, 3, 2), "cm"), axis.title.x=element_text(vjust=1)) + 
  geom_hline(yintercept = 0.73, color="red") + 
  xlab("Genome") + ylab("R-squared") + labs(title = "Drosophila simulans models R-squared values")
plot(rsq.plot)
pdf(file = self.dir.folder("rsq.drosoph.pdf"), width = 15, height=8)
plot(rsq.plot)
dev.off()
```

```{r drosophila clusters}
#first prepare the functions
source("R/calculate.clustering.table.R")
source("R/convert.core.list.matrix.R")
source("R/extract.cores.R")
#load the core-outsider data
core.distribution.data<- convert.core.list.matrix(clustering.list = calculate.clustering.table(
  bh.data = bh.data, rsquared.cutoff = 0.73, rsquared.table = model.df, legend = legend, processors = 4), legend = legend)
rsq.bar<- core.distribution.data %>% select.genomes(df=., g.ids=6) %>% mutate(id=rownames(.)) %>% melt(id.vars="id") %>% group_by(variable,value) %>%
  summarise(times=n()) %>% do({
    .$times<- rescale(.$times, from = c(0,sum(.$times)),to = c(0,1))
    return(.)
  }) %>% ungroup() %>% mutate(variable=str_replace_all(string=.$variable, pattern ="X", replacement="")) %>%
  merge(x=., y=legend, by.x="variable", by.y="id_genomes") %>% 
  merge(x=., y=filter(model.df,query.genome==6|target.genome==6) %>% mutate(variable=sapply(1:nrow(.),function(x){
    if(.$target.genome[x]==6){
      return(.$query.genome[x])
    }else{
      return(.$target.genome[x])
    }
  }))) %>% arrange(rsquared) %>% 
  mutate(name=str_trim(toupper(str_replace_all(string=.$name, pattern = "\\.fasta|_|complete_genome|chromosome", replacement=" ")))) %>%
  mutate(name=factor(.$name, levels=unique(.$name))) %>%
  ggplot(data=., mapping=aes(x=name, y=times, group=variable, fill=value, order=value)) +geom_bar(stat="identity") + 
  theme_bw() + theme(axis.text.x=element_text(angle=45, hjust=1),plot.margin = unit(c(1, 1, 1, 4), "cm")) +
  xlab("Genome") + ylab("Core/Outsider Ratio") + labs(title="Drosophila Core/Outsider Gene Classification") +
  geom_segment(mapping=aes(x=1, xend=nrow(legend)-1, y=0.1, yend=0.1), arrow=arrow(length=unit(0.5,"cm")),colour="red") +
  scale_fill_manual(values=c("#06960B","#999999","#F7728A"),labels=c("Core Genes","No Clear Separation","Outsider Genes"), guide=guide_legend(title=""))
plot(rsq.bar)
pdf(file=self.dir.folder("core.outsider.barplot.pdf"),width=15, height=6)
plot(rsq.bar)
dev.off()
```

Now we are going to use part of the above code to plot out an idealized tree.

```{r logistic regression cleaned tree}
#select only core genes
bh.data.cores<- bh.data[extract.cores(bh.data = bh.data, core.number.cutoff=5, core.outsider.df = core.distribution.data, processors = 6),]
#extract mle gprimes
bh.data.cores.mles.gprimes<- extract.gprimes(mle.list = getMLEs(df = expand.df(bh.data.cores)))
#distance
bh.data.cores.mles.gprimes.dist<- dist(bh.data.cores.mles.gprimes)
#fitch
rf<- Rfitch(D=bh.data.cores.mles.gprimes.dist,path = "/usr/local/bin", model="FM", power=2, negative=TRUE, global=FALSE,random.order=FALSE, root=FALSE, quiet=TRUE)
#normalize the genome names
rf.names<- tidy.tree(tree = rf, legend = legend)
output.file<- self.dir.folder("clean.tree.0.73.rsq.5.corecut.newick")
write.tree(phy = rf.names, file = output.file)
plot.itol(newick.file = output.file, output.file = paste0(output.file,".pdf"))
plot.itol(newick.file = output.file, output.file = paste0(output.file,".eps"), format = "eps", font.size = 120)
```

```{r drosophile wolbachia plus minus outsiders}
gp<- gen.pca(df = select.genomes(df = bh.data, g.ids = c(3,6)), legend = legend, g.ids = c(3, 6), circle = TRUE)
gp<- gp + theme_bw() + theme(legend.position="bottom")+ylim(-1,2.5) +xlim(-2.1, 0.8)
plot(gp)
pdf(file=self.dir.folder("drosophila.wolbachia.dirty.pdf"), width=15, height=15)
plot(gp)
dev.off()
#get rid of the uncertain genes
gp$data<- gp$data[extract.cores(bh.data = bh.data, core.outsider.df = core.distribution.data, processors = 6, core.number.cutoff = 5),]
plot(gp)
pdf(file=self.dir.folder("drosophila.wolbachia.clean.pdf"), width=15, height=15)
plot(gp)
dev.off()
```

Now i need a plot for the full core/outsider/nonsep distribution.

```{r full core distribution plot}
#extract the genomes
bar.plot.core.outsider<- expand.df(short.df = core.distribution.data) %>% select(-QUERY_ORF_ID) %>% 
  merge(x=., y=legend, by.x="ID_QUERY_GENOME", by.y="id_genomes")%>% select(-ID_QUERY_GENOME) %>%
  melt(id.vars="name") %>% group_by(name,value) %>% summarize(total=n()) %>% ungroup() %>% group_by(name) %>%
  do({
    .$total=rescale(.$total, from=c(0, sum(.$total)), to=c(0,1))
    return(.)
  }) %>% ungroup() %>% mutate(name=sapply(name, function(x){
    if(x=="WOLBACHIA ENDOSYMBIONT OF DROSOPHILA SIMULANS WHA"){
      return("WOLBACHIA DROSOPHILA")
    }
    if(x=="WOLBACHIA ENDOSYMBIONT OF CULEX QUINQUEFASCIATUS PEL"){
      return("WOLBACHIA CULEX")
    }
    return(x)
  })) %>% 
  mutate(name=sapply(name,function(x){
    x<- strsplit(x,split = " ",fixed = TRUE)
    return(paste(x[[1]][1],x[[1]][2],sep=" "))
    })) %>% filter(!grepl(name, pattern="ICHTHYOPHTHIRIU", fixed = TRUE))%>%
  ggplot(data=., mapping=aes(x=value, y=total, group=name, fill=name)) + geom_bar(stat="identity", position="dodge") +
  theme_bw() + theme(axis.text.x=element_text(angle=45, hjust=1, size=24), legend.position="none", title= element_text(size=32))+
  facet_wrap(~name,ncol=6) + xlab("") + ylab("proportion") + labs(title="Core/Outsider/NoSep Distribution Over Genomes")
plot(bar.plot.core.outsider)
pdf(file=self.dir.folder("bar.plot.core.outsider.pdf"), width=15, height=10)
plot(bar.plot.core.outsider)
dev.off()
```

```{r all genes distribution}
total.all<- (core.distribution.data %>% expand.df() %>% group_by(ID_QUERY_GENOME) %>% summarize(total.all=n()))$total.all
hist.plot<- core.distribution.data[extract.cores(bh.data = bh.data, core.number.cutoff=5, core.outsider.df = core.distribution.data, processors = 6),] %>%
  expand.df() %>% group_by(ID_QUERY_GENOME) %>% summarize(total.core=n()) %>% ungroup() %>% 
  mutate(total.all=total.all) %>% mutate(total.outsider=total.all-total.core) %>% 
  rename(core=total.core, outsider=total.outsider) %>%
  melt(id.vars="ID_QUERY_GENOME") %>% 
  group_by(ID_QUERY_GENOME) %>% 
  do({
    .$value<- rescale_max(.$value)
    return(.)
  }) %>% ungroup() %>% filter(variable%in%c("core","outsider")) %>%
  merge(x=., y=legend, by.y="id_genomes", by.x="ID_QUERY_GENOME") %>%
  mutate(name=sapply(name, function(x){
    if(x=="WOLBACHIA ENDOSYMBIONT OF DROSOPHILA SIMULANS WHA"){
      return("WOLBACHIA DROSOPHILA")
    }
    if(x=="WOLBACHIA ENDOSYMBIONT OF CULEX QUINQUEFASCIATUS PEL"){
      return("WOLBACHIA CULEX")
    }
    return(x)
  })) %>% 
  mutate(name=sapply(name,function(x){
    x<- strsplit(x,split = " ",fixed = TRUE)
    return(paste(x[[1]][1],x[[1]][2],sep=" "))
    })) %>% 
  filter(!grepl(name, pattern="ICHTHYOPHTHIRIU", fixed = TRUE))%>%
  ggplot(data=., mapping=aes(x=variable, y=value, fill=name)) +geom_bar(stat="identity", position="dodge") +
  facet_wrap(~name, ncol=5) + theme_bw() + 
  theme(axis.text.x=element_text(angle=45, hjust=1, size=24), legend.position="none", axis.title=element_text(size=28),plot.title=element_text(size=34)) +
  xlab("") + ylab("ratio") + labs(title="Core/Outsider Ratio Distribution Over Genomes")
plot(hist.plot)
pdf(file=self.dir.folder("bar.plot.core.outsider.pdf"), width=13, height=10)
plot(hist.plot)
dev.off()
``` 















